torch/nn/init.py

```diff
@@ -1,239 +1,394 @@
+        with torch.no_grad():
+            return tensor.normal_(self.mean, self.std)
 
-    return fan_in, fan_out
 
+class Dirac(Initializer):
+    r"""Fills the {3, 4, 5}-dimensional input `Tensor` with the Dirac
+    delta function. Preserves the identity of the inputs in `Convolutional`
+    layers, where as many input channels are preserved as possible.
+
+    Args:
+        tensor: a {3, 4, 5}-dimensional `torch.Tensor`
 
-def xavier_uniform(tensor, gain=1):
-    """Fills the input Tensor or Variable with values according to the method
+    Examples:
+        >>> w = torch.Tensor(3, 16, 5, 5)
+        >>> nn.init.dirac_(w)
+        >>> dirac_init = nn.init.dirac_()
+        >>> dirac_init(w)
+    """
+    def __call__(self, tensor):
+        dimensions = tensor.ndimension()
+        if dimensions not in [3, 4, 5]:
+            raise ValueError("Only tensors with 3, 4, or 5 dimensions are supported")
+
+        sizes = tensor.size()
+        min_dim = min(sizes[0], sizes[1])
+        with torch.no_grad():
+            tensor.zero_()
+
+            for d in range(min_dim):
+                if dimensions == 3:  # Temporal convolution
+                    tensor[d, d, tensor.size(2) // 2] = 1
+                elif dimensions == 4:  # Spatial convolution
+                    tensor[d, d, tensor.size(2) // 2, tensor.size(3) // 2] = 1
+                else:  # Volumetric convolution
+                    tensor[d, d, tensor.size(2) // 2, tensor.size(3) // 2, tensor.size(4) // 2] = 1
+        return tensor
+
+
+class Sparse(Initializer):
+    r"""Fills the 2D input `Tensor` as a sparse matrix, where the
+    non-zero elements will be drawn from the normal distribution
+    :math:`\mathcal{N}(0, 0.01)`, as described in "Deep learning via
+    Hessian-free optimization" - Martens, J. (2010).
+
+    Args:
+        tensor: an n-dimensional `torch.Tensor`
+        sparsity: The fraction of elements in each column to be set to zero
+        std: the standard deviation of the normal distribution used to generate
+            the non-zero values
+
+    Examples:
+        >>> w = torch.Tensor(3, 5)
+        >>> nn.init.sparse_(w, sparsity=0.1)
+        >>> sparse_init = nn.init.sparse_(sparsity=0.1)
+        >>> sparse_init(w)
+    """
+    def __init__(self, sparsity, std=0.01):
+        self.sparsity = sparsity
+        self.std = std
+
+    def __call__(self, tensor):
+        if tensor.ndimension() != 2:
+            raise ValueError("Only tensors with 2 dimensions are supported")
+
+        rows, cols = tensor.shape
+        num_zeros = int(math.ceil(rows * self.sparsity))
+
+        with torch.no_grad():
+            tensor.normal_(0, self.std)
+            for col_idx in range(cols):
+                row_indices = list(range(rows))
+                random.shuffle(row_indices)
+                zero_indices = row_indices[:num_zeros]
+                for row_idx in zero_indices:
+                    tensor[row_idx, col_idx] = 0
```

Comment:
Could you rebase on top of master? A PR that optimized this operation has been recently merged